tests/suite/nginxgateway_test.go

```diff
@@ -0,0 +1,180 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	core "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+
+	ngfAPI "github.com/nginxinc/nginx-gateway-fabric/apis/v1alpha1"
+	"github.com/nginxinc/nginx-gateway-fabric/tests/framework"
+)
+
+var _ = Describe("NginxGateway", Ordered, Label("functional", "nginxGateway"), func() {
+	var (
+		ngfPodName string
+
+		namespace          = "nginx-gateway"
+		nginxGatewayNsname = types.NamespacedName{Name: releaseName + "-config", Namespace: namespace}
+
+		files = []string{
+			"nginxgateway/nginx-gateway.yaml",
+		}
+	)
+
+	AfterAll(func() {
+		// cleanup and restore NGF instance
+		teardown(releaseName)
+		setup(getDefaultSetupCfg())
+	})
+
+	getNginxGateway := func(nsname types.NamespacedName) (ngfAPI.NginxGateway, error) {
+		ctx, cancel := context.WithTimeout(context.Background(), timeoutConfig.GetTimeout)
+		defer cancel()
+
+		var nginxGateway ngfAPI.NginxGateway
+
+		if err := k8sClient.Get(ctx, nsname, &nginxGateway); err != nil {
+			return nginxGateway, err
+		}
+
+		return nginxGateway, nil
+	}
+
+	verifyAndReturnNginxGateway := func(nsname types.NamespacedName) ngfAPI.NginxGateway {
+		nginxGateway, err := getNginxGateway(nsname)
+		Expect(err).ToNot(HaveOccurred())
+		Expect(nginxGateway).ToNot(BeNil())
+
+		Expect(nginxGateway.Status.Conditions).To(HaveLen(1))
+		condition := nginxGateway.Status.Conditions[0]
+
+		Expect(condition.Type).To(Equal("Valid"))
+		Expect(condition.Reason).To(Equal("Valid"))
+
+		return nginxGateway
+	}
+
+	getNGFPodName := func() (string, error) {
+		podNames, err := framework.GetReadyNGFPodNames(
+			k8sClient,
+			ngfNamespace,
+			releaseName,
+			timeoutConfig.GetTimeout,
+		)
+		if err != nil {
+			return "", err
+		}
+
+		if len(podNames) != 1 {
+			return "", fmt.Errorf("expected 1 pod name, got %d", len(podNames))
+		}
+
+		return podNames[0], nil
+	}
+
+	When("testing NGF on startup", func() {
+		When("log level is set to debug", func() {
+			It("outputs debug logs and the status is accepted and true", func() {
+				ngfPodName, err := getNGFPodName()
+				Expect(err).ToNot(HaveOccurred())
+
+				_ = verifyAndReturnNginxGateway(nginxGatewayNsname)
+
+				logs, err := resourceManager.GetPodLogs(ngfNamespace, ngfPodName, &core.PodLogOptions{
+					Container: "nginx-gateway",
+				})
+				Expect(err).ToNot(HaveOccurred())
+
+				Expect(logs).To(ContainSubstring("\"level\":\"debug\""))
+			})
+		})
+
+		When("default log level is used", func() {
+			It("only outputs info logs and the status is accepted and true", func() {
+				teardown(releaseName)
+
+				cfg := getDefaultSetupCfg()
+				cfg.infoLogLevel = true
+				setup(cfg)
+
+				ngfPodName, err := getNGFPodName()
+				Expect(err).ToNot(HaveOccurred())
+
+				nginxGateway := verifyAndReturnNginxGateway(nginxGatewayNsname)
```

Comment:
I'm worried that if any previous functional test messes with the observed generation of the nginxGateway, this test will fail since we didn't start at `1`. I got around this previously by grabbing the current observed generation of the nginxGateway and basing the test assertion off of that. However, I can't really do that anymore since the observed generation has to be specified "as" the nginxGateway is being verified.

I made the change and maybe you can see why I'm worried too. But the test works.

Reply:
> I'm worried that if any previous functional test messes with the observed generation of the nginxGateway, this test will fail since we didn't start at 1.

This shouldn't happen since the observed gen is only incremented when updates are made to the NginxGateway and other tests shouldn't be updating the NginxGateway resource.

> I got around this previously by grabbing the current observed generation of the nginxGateway and basing the test assertion off of that. However, I can't really do that anymore since the observed generation has to be specified "as" the nginxGateway is being verified.

You can still do that when you update the NginxGateway and check that the observed generation is incremented:

```
gen := getCurrentNginxGatewayObservedGeneration(nsname)
verifyNginxGatewayStatus(nsname, gen +1)
```

or, so we don't GET the NginxGateway twice:

```
ng := getNginxGateway(nsname)
gen := getCurrentObservedGeneration(ng)
verifyNginxGatewayStatus(ng, gen+1)
```

We could also take a leaf out of the Gateway API's conformance tests and just check that the observed generation is the same as the object's generation: https://github.com/kubernetes-sigs/gateway-api/blob/24f73922acf501b547e0f183c8f8066eb6fbb045/conformance/utils/kubernetes/helpers.go#L160

Reply:
Made a change, think it fixes it. Checked out the Gateway API's tests, but did not incorporate it, though it was useful looking at it.